### ISSUE TITLE
Systematic support for checked and synthesised levels

### ISSUE THREAD

**Issue by laMudri** [agda/agda-stdlib#1339]
I thought about replying to #312, but I think the angle I'm coming at is sufficiently different to justify a new issue.

There are two coherent ways to manage universe levels, corresponding to the two directions of bidirectional type checking:

1. Every construct is given a canonical level, and constructs containing other constructs have their level calculated based on the levels of their sub-constructs, combined with `0ℓ` and `_⊔_`. (synthesising)
1. We make sure we always know the level of the thing we're constructing beforehand, and the levels of any constituent parts are based on the level of the whole construct, usually being equal. (checking)

Both directions can be useful, though Agda has slightly better support for the former. Therefore, stdlib should export definitions fitting both styles, ideally in a systematic way.

Most of stdlib, for example `Data.Unit` and `Data.Product`, supports the synthesising style (I guess this followed from Agda's slightly better support for it). In `Data.Unit`, `⊤` knows its level (`0ℓ`), and in `Data.Product`, the level of `A × B` is calculated from the levels of `A` and `B` (`a ⊔ b`). But we now have a few modules like `Data.Unit.Polymorphic`, which fit the checking style – the only way to know what level `⊤` from `Data.Unit.Polymorphic` should have is to check `⊤` with the level already known (and whatever level that is, it will work). `Data.Unit.Polymorphic` is incompatible with `Data.Product`, in the sense that `⊤ × A` is in most cases going to come up yellow. Therefore, I think we should, analogously to `Data.Unit.Polymorphic`, have a version of `_×_` with type `∀ {ℓ} (A : Set ℓ) (B : Set ℓ) → Set ℓ`.

It turns out, for these simple connectives, that either the synthesising variant can be implemented as a specialisation of the checking variant or vice versa. I claim this erroneously leads to us considering the more general one the “right one”, when in fact we need both. We could, but for largely historical reasons currently don't, implement the `⊤` of `Data.Unit` as a specialisation of the `⊤` of `Data.Unit.Polymorphic`:

```agda
module Data.Unit.Polymorphic where

  record ⊤ {ℓ} : Set ℓ where
    constructor tt

module Data.Unit where

  import Data.Unit.Polymorphic as Chk
  open import Level

  ⊤ : Set 0ℓ
  ⊤ = Chk.⊤

  open Chk public using (tt)
```

Conversely, we could implement a *checking* variant of `Data.Product` in terms of the normal (synthesising) one:

```agda
module Data.Product.Checking where

  import Data.Product as Syn

  Σ : ∀ {ℓ} (A : Set ℓ) (B : A → Set ℓ) → Set ℓ
  Σ = Syn.Σ

  _×_ : ∀ {ℓ} (A B : Set ℓ) → Set ℓ
  A × B = Σ A λ _ → B

  -- Reimplement any other types too, but functions are fine...

  open Syn public hiding (Σ; _×_)
```

I don't have a good grasp as to why one is sometimes more general than the other, but it seems like an easy thing to take advantage of. Everything I've said here can apply to more complex things, like algebraic constructs, where the sharing of definitions becomes yet more useful.

To summarise, here are some proposals, with varying levels of importance and feasibility:

1. Rename `.Polymorphic` modules to `.Checking`.
2. Introduce `Data.Product.Checking`, `Data.Sum.Checking`, &c, as outlined above.
3. Gut `Data.Unit` and `Data.Empty`, defining them via the `.Polymorphic`/`.Checking` variants.
4. Provide both variants for algebraic constructs (e.g, `Algebra.Construct.Zero`, `Algebra.Construct.DirectProduct`, `Relation.Binary.Construct.Always`), with the same module layout as the data types use.
5. Justify these choices in a README (related to #312).

**Comment by jespercockx** [comment]
Do you have examples of Agda code written in 'checking' style for the levels?

**Comment by gallais** [comment]
> Do you have examples of Agda code written in 'checking' style for the levels?

This is what we do in agdarsec to some extent: a parser is parametrised
by a level `l` and everything it produces & manipulate is below that level.
Instead of going down the road James suggests, we have some [library
support](https://github.com/gallais/agdarsec/blob/master/src/Level/Bounded.agda) for bounded level quantification.

It works well for our purposes but then again we only ever program using
these, we never prove anything about the combinators themselves. I have
no idea whether this approach would be tractable if you also wanted to
prove things about the functions you define.
